### PR TITLE
Add CKB wallet functions

### PR DESCRIPTION
**Pull request by matthewbauer**
This is an initial drafts of CKB wallet functions. It adds:

- "getAppConfiguration"
- "getWalletId"
- "getWalletPublicKey"
- "signTransaction"

These are pretty simple and should be self explanatory. They map to
APDUs on the nervos ledger app available here: https://github.com/obsidiansystems/ledger-app-nervos.

### CODE REVIEW

File: packages/hw-app-ckb/src/Ckb.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+   * @param rawTxHex transaction to sign
+   * @param contextTransaction transaction context to use in parsing (optional)
+   * @return a signature as hex string
+   * @example
+   * const signature = await ckb.signTransaction("44'/144'/0'/0/0", "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000");
+   */
+  async signTransaction(
+    path: string,
+    rawTxHex: string,
+    rawContextTxHex: string
```
Comment (jonored) [inline packages/hw-app-ckb/src/Ckb.js]: I think this needs to be an array; we have one context transaction per input in the current transaction.

File: packages/hw-app-ckb/src/Ckb.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    let rawPath = Buffer.alloc(1 + bipPath.length * 4);
+    rawPath.writeInt8(bipPath.length, 0);
+    bipPath.forEach((segment, index) => {
+      rawPath.writeUInt32BE(segment, 1 + index * 4);
+    });
+    await this.transport.send(0x80, 0x03, 0x00, 0x00, rawPath);
+
+    const maxApduSize = 230;
+
+    if (rawContextTxHex) {
```
Comment (jonored) [inline packages/hw-app-ckb/src/Ckb.js]: And then we need to loop over context transactions here.

**Comment by gre** [comment]
@matthewbauer tests seem to fail on CI btw. seems to be wrong apdu replay

if you could take a look and fix.
thanks

File: packages/hw-app-ckb/src/Ckb.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+      let rawContextTx =
+        rawContextTxHex !== null ? Buffer.from(rawTxHex, "hex") : null;
+      for (let i = 0; i < Math.floor(rawContextTx.length / maxApduSize); i++) {
+        let data = Buffer.alloc(maxApduSize);
+        for (let j = 0; j < maxApduSize; j++) {
+          let index = i * maxApduSize + j;
+          if (index >= rawContextTx.length) {
+            break;
+          }
+          rawContextTx.copy(data, 1, index);
```
Comment (gre) [inline packages/hw-app-ckb/src/Ckb.js]: we seem to do byte per byte, it's not possible to do it in a chunk of byte with copy? apologies if i miss the big picture

File: packages/hw-app-ckb/README.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
+[Github](https://github.com/LedgerHQ/ledgerjs/),
+[Ledger Devs Slack](https://ledger-dev.slack.com/)
+
+## @ledgerhq/hw-app-ckb
+
+Ledger Hardware Wallet CKB JavaScript bindings.
+
+## Using LedgerJS for XTZ
```
[the remark this answers is not in the document]
Comment (jonored) [reply packages/hw-app-ckb/README.md]: Yep, fixed.

File: packages/hw-app-ckb/README.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        console.log(signature);
+    };
+
+    doAll().catch(err => console.log(err));
+
+To get this to work with a dev checkout of the LedgerJS repo, make sure that
+you have appropriate dependencies in your `package.json`, either by using
+`yarn add file:.../path` or `yarn link`. Make sure that you `yarn install`
+_and_ `yarn build` the LedgerJS packages, as `yarn build` will be necessary
+to translate the Javascript into a form acceptable by Node.
```
[the remark this answers is not in the document]
Comment (jonored) [reply packages/hw-app-ckb/README.md]: Removed.

File: packages/hw-app-ckb/README.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+If you want to use web instead, replace `hw-transport-node-hid` with `hw-transport-u2f`
+and run `npx webpack`.
```
[the remark this answers is not in the document]
Comment (jonored) [reply packages/hw-app-ckb/README.md]: Also removed.

File: packages/hw-app-ckb/src/Ckb.js [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    await this.transport.send(0x80, 0x03, 0x00, 0x00, rawPath);
+
+    const maxApduSize = 230;
+
+    if (!rawContextsTxHex) rawContextsTxHex = [];
+
+    for (const rawContextTxHex of rawContextsTxHex) {
+      let rawContextTx =
+        rawContextTxHex !== null ? Buffer.from(rawTxHex, "hex") : null;
+      for (let i = 0; i < Math.floor(rawContextTx.length / maxApduSize); i++) {
```
[the remark this answers is not in the document]
Comment (jonored) [reply packages/hw-app-ckb/src/Ckb.js]: Done.

File: packages/hw-app-ckb/src/Ckb.js [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+      let lastContextData = Buffer.alloc(maxApduSize);
+      let lastContextOffset =
+        Math.floor(rawTx.length / maxApduSize) * maxApduSize;
+      for (
+        let index = lastContextOffset;
+        index < rawContextTx.length;
+        index++
+      ) {
+        rawContextTx.copy(lastData, 1, index);
```
[the remark this answers is not in the document]
Comment (jonored) [reply packages/hw-app-ckb/src/Ckb.js]: Yes, this is fixed to not even perform a copy now.

File: packages/hw-app-ckb/src/Ckb.js (hunk 3 above)
Comment (jonored) [reply packages/hw-app-ckb/src/Ckb.js]: Fixed and the copy removed, even; it was not required.

**Comment by matthewbauer** [comment]
Closing. We're now going to maintain hw-app-ckb, available here:

https://github.com/obsidiansystems/hw-app-ckb

This is the same as this PR, but with just the Nervos wallet function part.

**Pull request closed without merging**